File: 1214-Two Sum BSTs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        
# first try
        arr1 = []
        def dfs(node, arr):
            if not node:
                return
            
            dfs(node.left, arr)
            arr.append(node.val)
            dfs(node.right, arr)
        
        dfs(root1, arr1)
        
        def dfs2(node):
            if not node:
                return False
            
            if dfs2(node.left):
                return True
            idx = bisect_left(arr1, target - node.val)
            if idx < len(arr1) and target - arr1[idx] == node.val:
                return True
            if dfs2(node.right):
                return True
            return False
        
        return dfs2(root2)
            
# very brilliant solution, use in-order and reverse in-order iterator for BST
        stack1, stack2 = [], []
    
        while True:
            while root1:
                stack1.append(root1)
                root1 = root1.left
            while root2:
                stack2.append(root2)
                root2 = root2.right
            if not stack1 or not stack2:
                return False
            
            currVal = stack1[-1].val + stack2[-1].val
            
            if currVal == target:
                return True
            elif currVal < target:
                root1 = stack1.pop().right
            else:
                root2 = stack2.pop().left

File: C++/1214-Two Sum BSTs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool twoSumBSTs(TreeNode* root1, TreeNode* root2, int target) {
        stack<TreeNode*> stk1, stk2;
        
        while (true){
            while (root1){
                stk1.push(root1);
                root1 = root1->left;
            }
            while (root2){
                stk2.push(root2);
                root2 = root2->right;
            }
            if (stk1.empty() || stk2.empty() ) return false;
            
            int currSum = stk1.top()->val + stk2.top()->val;

            if (currSum == target)
                return true;
            else if (currSum < target){
                root1 = stk1.top();
                stk1.pop();
                root1 = root1->right;               
            }
            else{
                root2 = stk2.top();
                stk2.pop();
                root2 = root2->left;                
            }
        }
    }
};


// Redo

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool twoSumBSTs(TreeNode* root1, TreeNode* root2, int target) {
        stack<TreeNode*> st1, st2;
        while (root2 != nullptr) {
            st2.push(root2);
            root2 = root2->right;
        }

        while (true) {
            if (root1 != nullptr) {
                while (root1 != nullptr) {
                    st1.push(root1);
                    root1 = root1->left;
                }
            } else if (!st1.empty() && !st2.empty()) {
                root1 = st1.top();
                st1.pop();

                while (root1->val + st2.top()->val >= target) {
                    root2 = st2.top();
                    st2.pop();
                    if (root1->val + root2->val == target) return true;
                    else if (root2->left != nullptr) {
                        root2 = root2->left;
                        while (root2 != nullptr) {
                            st2.push(root2);
                            root2 = root2->right;
                        }
                    } else if (st2.empty()) {
                        break;
                    }
                }

                root1 = root1->right;
            } else {
                break;
            }
        }

        return false;
    }
};